### ISSUE TITLE
Invalid suggestion (type incorrect)

### ISSUE THREAD

**Issue by LeventErkok** [ndmitchell/hlint#980]
This started happening recently:

```
$ cat a.hs
infixl 8 ^.
a ^. b = error "undefined"

t a = a ^. 1 ^. 1
```

```
$ hlint --version
HLint v2.2.11, (C) Neil Mitchell 2006-2020
$ hlint a.hs
a.hs:4:11: Warning: Use .
Found:
  a ^. 1 ^. 1
Perhaps:
  a ^. (1 . 1)

1 hint
```
The suggestion doesn't type check.

Perhaps its a breakage in an upstream tool? (Looks like maybe the parser is doing something funky that throws hlint off.) Happy to provide more details.

**Comment by googleson78** [comment]
The default `lens` group is "at fault":
```yaml
- group:
    name: lens
    enabled: true
    imports:
    - package base
    - package lens
    rules:
    - warn: {lhs: "(a ^. b) ^. c", rhs: "a ^. (b . c)"}
```

**Comment by googleson78** [comment]
Ok... wat:
```haskell
blademaster :: ~ » cat a.hs
infixl 8 ^.
a ^. b = error "undefined"

t a = a ^. 1 ^. 1
```

With only the group:
```yaml
blademaster :: ~ » cat hlint.yaml
- group:
    name: lens
    enabled: true
    imports:
    - package base
blademaster :: ~ » hlint -h hlint.yaml a.hs
a.hs:4:7-17: Warning: Redundant ^.
Found:
  a ^. 1 ^. 1
Perhaps:
  a ^. (1 . 1)

1 hint
```

but disabling the group disables the hint!:
```yaml
blademaster :: ~ » cat hlint.yaml
- group:
    name: lens
    enabled: false
    imports:
    - package base
blademaster :: ~ » hlint -h hlint.yaml a.hs
No hints
```

**Comment by ndmitchell** [comment]
So the issue here is that `^.` is a lens operator we hint about, and we currently don't do a fantastic job about tracking the provenance of an identifier, so we assume `^.` is probably a lens operator, and hint appropriate. The ultimate long-term solution is to figure out the provenance of operators better, but that's a big job. In the meantime we either accept that other users of `^.` might be tripped up, or disable that lens hint. Thoughts?

**Comment by LeventErkok** [comment]
I see.. But looks like hlint *does* keep track of the provenance for functions. Is that correct?

What's the difference between an operator and a function in this context, i.e., as far as hlint is concerned?

**Comment by ndmitchell** [comment]
Functions and operators are the same - HLint uses heuristics to guess the provenance of both, and in both cases, it's pretty dubious.

**Comment by LeventErkok** [comment]
I see.  I was able to work around this by putting a module-level exception into the file and it seems to do the trick.

I really don't have a strong opinion on whether you should disable this hint or not; it sure seems useful. But perhaps hlint can put a "warning" saying it's assuming this is the lens version of `^.` whenever it isn't absolutely sure? If you can do that, it'd help the end-user at least see where the problem is coming from.

I'll leave this ticket open, but feel free to close it as you see fit!

**Comment by ndmitchell** [comment]
Since HLint is only roughly sure for pretty much every identifier, you'd have to go as far as saying that everywhere, which would somewhat defeat the point of HLint. I'll leave this ticket open for the moment, then go through and synthesise the tickets where matching gets it a bit wrong, so we have a complete understanding of the problem.
